### DISCUSSION TITLE
Magick++ NoDecodeDelegateForThisImageFormat `PNG' @ error/constitute.c/ReadImage/572 | VisualDynamicMT

### DISCUSSION THREAD

**Discussion by oblivioncth** [ImageMagick/ImageMagick#4238] [Help]
Hi,

Sorry if this is the wrong place, but I'm new to using Discussions on Github and wasn't expecting the old forum to be retired.

Anyway, I compiled ImageMagick from source so that I could more easily debug an application using Magick++/MagickWand since it crashes in the Debug build when using the prebuilt release binaries due to read violations when Magick functions are called (no surprise there). 

Setup:
Compiler - MSVC2019 (Toolset v142/WDK 10.0.18362.0)
Target Arch - x64
Link Type - VisualDynamicMT

I built using the steps provided for Windows here on GitHub (as well as the ones on the site) which went fine after solving a few hiccups.  The application I have is linked against CORE_DB_Magick++_.lib, CORE_DB_MagickCore_.lib, and CORE_DB_MagickWand_.lib and runs fine once the DLLs for those libs and a good number of other DLLs for ImageMagick's own dependencies are present aside my target output. The issue is that as soon the input image is to be read this exception given in the title is thrown.

The application does not have this issue if the latest prebuilt release package for Windows is installed and is pointed at its libs/includes instead; however, the issue is that this application isn't my own and is one I'm trying to study so that I can partially port its implementation into another app. It includes a CMake script that generates a VisualStudio solution after locating all of the dependencies on your system, including an installed release of ImageMagick, which as I mentioned it works with. Because of this I'm not entirely sure how exactly the project gets configured to use the ImageMagick libs once its an .sln, outside of the obvious C++ include and Linker input fields.

When I change the the project to use my own compiled ImageMagick in the Debug build I'm simply adding the above libraries to the Linker input section (along with its parent folder to the Additional Libraries section), as well as the includes under C++/Additional Includes. I'm guessing there must be some difference in how the project needs to be configured, or something that needs to be done in code to use the Debug libraries correctly for the build in delegates to work right. but I don't know what as this app is my first exposure to the ImageMagick API.
 
Convert.exe built from UTIL_convert in my local VS build of the libraries in Debug doesn't have this issue though. It lists PNG as a delegate with the -version switch and I confirmed it was able to convert a PNG to a JPEG. This indicates to me that the build of ImageMagick itself is fine, and that again, I have something misconfigured in my target application.

Hoping someone can point me in the right direction.

Thanks.

**Comment by snibgo** [comment]
You don't mention that you have built the delegate libraries, such as libpng. If you haven't done that, then that is (part of) the problem.

**Reply by oblivioncth** [reply]
I see how this makes sense in theory, but I find this a bit confusing.

Pretty much all of the info that can be found for building delegates reference to ImageMagick's use on Linux where it seems like the more official packages may not come with binaries for the delegates by default since its easy enough to builds/build from source through a package manager, but there is nearly nothing for Windows. Not even the "Advanced Windows Installation" page here talks about their existence, let alone building them: https://imagemagick.org/script/advanced-windows-installation.php.

I know they are all separate, individual libs that aren't associated with ImageMagick but I'd figure there would be some documentation on how to use them correctly with ImageMagic (in the shared case I guess it dynamically looks for the DLLs?) on Windows. There is some info in the "Advanced Linux Installation" page, but obviously much of that doesn't apply outside of that platform.

Additionally, if this is the case, then how does `convert.exe` have all of the standard delegates (including PNG) included (I'm guessing statically linked since they're marked as built-in) and is able to successfully convert PNG -> JPG? I assume(d) that it obtained the libs from the `CloneRepositories.cmd` step in the Windows build process. Is there some API code I can execute to get a list of available delegates within the currently initialized library instance, similar to to `convert -version`? Maybe this way I could at least confirm that list is empty (or at least missing PNG).

I do wonder if the issue is that I do have binaries of these libraries (due to the above noted step) and that they are statically linked to the convert utility, but my application expects them dynamically due to its configuration.

I did make sure to use the configure tool to setup a build for DynamicMT VisualMagick but perhaps that doesn't actually also obtain dynamic builds of the third party libraries and so the build only works "out-of-the-box" if you use the static configuration? I wouldn't mind trying a static build of ImageMagick as well to see, but I got stuck compiling that due to 4 LNK2019's pertaining to `__declspec(dllimport) public: __cdecl Magick::qualityImage::qualityImage(unsigned __int64) `, though that's probably for another question.

Thoughts?

**Comment by snibgo** [comment]
> Pretty much all of the info that can be found for building delegates reference to ImageMagick's use on Linux where it seems like the more official packages may not come with binaries for the delegates by default since its easy enough to builds/build from source through a package manager, but there is nearly nothing for Windows.

I describe my Windows build process at [Compiling IM with Cygwin](http://im.snibgo.com/compim.htm).

In a nutshell: I download the large (132MB) zip file from https://imagemagick.org/script/install-source.php#windows. Unzip, then build IM (automake, autoconf, ./configure, make, make install). This makes a working ImageMagick with built-in delegates for PNG, JPEG, raw etc, and my own filters (aka "process modules"). I can also build my own programs that call IM functions to read images from files, process images and write them to files.

I don't use MSVC, and I don't know what has to be done to ensure that IM is build with internal delegates, and that your own programs can call IM and indirectly use those delegates.

In particular, you mention a "Debug build". I don't know what that is, or whether all libraries need to be built as Debug. Some posters in this forum have said that when they don't use Debug, everything just works.

**Reply by oblivioncth** [reply]
Ok after looking that over I can confirm that "I" am building the delegates. I don't know what the exact setup is since the ImageMagick configure tool takes care of creating the Visual Studio projects for you, but the ImageMagick-Windows.zip you mention is the alternate to the CloneRepositories.cmd option that is mentioned just before in those instructions. It takes care of cloning all of ImageMagick's dependencies from git (I think) instead of using the prepackaged ones in the zip. I ended up with all of the same folders that are present in the zip, including the png folder with its related projects (i.e. libpng, CORE_png, IM_MOD_png, etc.), as well as the same for the many other default formats. This makes sense given that the bin/convert.exe that gets built from the main project does have the standard delegates built in and can use them.

Though, it seems at some point I made some kind of mistake, as it turns out when I recalled convert.exe working correctly with the delegates I must have been thinking of the static build that I couldn't get working with the application, as suddenly the Dynamic build's convert tool made the same complaint about delegates missing. Well I reconfigured VisualMagick, recompiled, and I believe a few more binaries were created this time (namely CORE_png_.dll). I did select "Include all demo, test, appl, and contrib directories" this time around, though that doesn't sound like it would affect the binary libs...

Regardless, now the Dynamic build of convert works correctly and my application that links to it is also able to load its target PNG without complaining of missing delegates.

As for the "Debug" build, I think this is somewhat a Windows vs. Linux thing. As I understand it, many source packages of software that come with nice build scripts commonly use a "Release with Debugging Info" style build so that you can use the built binaries in everyday situations or step through them in a debugger no problem, making a pure Debug build often unnecessary and something that more Linux oriented peeps don't even think about about.

On Windows however, at least in my experience, that build configuration is quite uncommon, at least as a default that wasn't purposely setup by the end user, probably due to conventions that were established by how MSVC handles things. So, generally you have a strict Debug build and Strict Release build, where the latter is impossible to debug with whatsoever (baring debugging the machine code) due to high levels of optimization. With ReleaseWithDebugInfo builds the optimization is just low enough that mangling and restructuring aren't bad enough to keep the debug symbols from working with the original source.

In this case, I required a Debug build of IM so that I could build a Debug release of the application I want to study/step-through, because mixing a Debug application with Release libraries can often cause issues, and did here as you get access violation errors when IM tries to read any files.

So in the end I'm not 100% sure what I did, but I'd say to make sure that the CORE and IM_MOD binaries for the image types you need are being produced and that you have all of them copied into the folder where your target application lives so that they can be found.
